### PR TITLE
Deflake TestNoReadWhenOff

### PR DESCRIPTION
**Pull request by xacrimon**
Updates https://github.com/gravitational/teleport/issues/9492#issuecomment-1150334615

This test was flaky because the semaphore would be released as soon as the multiplexer read the data, continuing the test. But this could potentially allow the test to continue execution and enable the multiplexer before the multiplexer state is read internally to check if we should discard or keep the data, which would prevent the data from being discarded, providing garbage data to the verifying read at the end of the test. This was fixed by waiting on a callback that's only called from within the multiplexer when it has discarded some data.

This PR also fixes a bug where the state watcher goroutine would be leaked on a Read() call if the return case is error or EOF.

### CODE REVIEW

**Review by Joerger — APPROVED** [review]

File: lib/srv/termmanager.go [hunk 1]
```diff
@@ -126,6 +129,7 @@ func (g *TermManager) Read(p []byte) (int, error) {
 	}
 
 	q := make(chan struct{})
+	defer close(q)
```
Comment (strideynet) [inline lib/srv/termmanager.go:132]: Nice work tidying up this goroutine leak.

File: lib/srv/termmanager_test.go [hunk 2]
```diff
@@ -89,41 +88,43 @@ func TestBufferedKept(t *testing.T) {
...
 
-type funcReader func(p []byte) (n int, err error)
-
-func (f funcReader) Read(p []byte) (n int, err error) {
-	return f(p)
-}
-
 func TestNoReadWhenOff(t *testing.T) {
 	m := NewTermManager()
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*15)
```
Comment (strideynet) [inline lib/srv/termmanager_test.go:93]: nit: perhaps we could go with something more conservative than fifteen seconds or we could make this `t.Parallel()`. Having read the implementation of the underlying code, I know that this test will be extremely quick, but if I hadn't and was glancing at this test, I might wonder why a test that could run for 5-10 seconds isn't in parallel.

**Review by strideynet — APPROVED** [review]
I had a dive into this this morning to have a look if we could avoid modifying the underlying implementation, and also still deflake this. As far as I can tell, your solution is probably the best. 

The only other thing I could come up with was just inserting a `time.Sleep` for some short period, and whilst that's simpler, it's certainly not as clean and is likely to flake again in future if the implementation of TermManager gets more complicated.

File: lib/srv/termmanager_test.go (hunk 2 above)
Comment (xacrimon) [reply lib/srv/termmanager_test.go:94]: Marked as parallel.

**Comment by xacrimon** [comment]
@strideynet Yep, I was looking for the same type of solution but quickly came to the conclusion that the term manager is our only source of truth for what it does internally and that we need to expose notifications like this PR does. Sadly termmanager is quite complicated, I did rewrite it a couple times to try to come up with a better concurrency model but was largely unable to.

**Comment by xacrimon** [comment]
@zmb3 Could you review this as a code owner please?

**Review by zmb3 — APPROVED** [review]

**Pull request merged**